pal03377/Athena-New
# Add automatic evaluation with LLM-as-a-Judge, LangSmith export, and SGI evaluation

### Motivation and Context



We want an option to systematically evaluate the modules without manual labor. 

closes #84 
closes #36 

### Description


- [x] Add `/evaluation` endpoint to Athena for returning arbitrary evaluation results 
- [x] Add example of how this `/evaluation` endpoint would like to `module_example` (for programming exercises)
- [x] Add this new module request to the Playground
- [x] Run evaluation for all healthy modules of the same type that support evaluation in evaluation mode
- [x] Display progress of automatic evaluation (and move "Start Generating" button)
- [x] Add LLM-as-a-judge for text modules
- [x] Disable LLM-as-a-judge using environment variable
- [x] Add token usage and response times in evaluation response (coming from LangSmith)
- [x] Add eval code for SGI evaluation


**Note:** I omitted the "module_text_llm estimated Accepted with Minor modifications" text from the UI since it is kind of ugly to add and does not give immediate benefit. In the future one might make automatic results available for the inline feedback view but this depends also on future experiment designs. 

### Steps for Testing


1. Run an evaluation with `module_text_llm`
2. After everything is done click export, it should export the `automatic_evaluation` now as a separate file
3. File should contain `LLM-as-a-judge`, `llm_statistics` (from LangSmith, I added my key to the env), and `feedback_statistics`
4. Cancel experiment
5. Import results and automatic evaluation and check if it works

### Screenshots


**Docs:**
<img width="532" alt="image" src="https://github.com/ls1intum/Athena/assets/5898705/a81d0ad0-ed17-4d76-ad10-374e49acfbda">


#### Not Started
<img width="620" alt="Screenshot 2023-11-07 at 16 36 26" src="https://github.com/ls1intum/Athena/assets/5898705/eaab1eff-39b9-46d5-ba3c-655dce6ff512">

#### Sent Training Feedback
<img width="634" alt="Screenshot 2023-11-07 at 16 36 31" src="https://github.com/ls1intum/Athena/assets/5898705/860bdef0-e7dd-4f44-9ec7-3dc1c3bbfd23">

#### Generated Suggestions
<img width="624" alt="Screenshot 2023-11-07 at 16 38 37" src="https://github.com/ls1intum/Athena/assets/5898705/9abc3cb4-c8cd-4371-805f-b5c897ba5869">

#### Finished
<img width="622" alt="Screenshot 2023-11-07 at 16 38 42" src="https://github.com/ls1intum/Athena/assets/5898705/a1ff49b1-0822-464f-8e8d-665ca85d4b14">
